File: solutions/ocaml/060-validate-binary-search-tree.ml
(*
  Problem 60: Validate Binary Search Tree (LeetCode 98)
  Difficulty: Med
  Language: OCaml
*)
type tree = Nil | Node of int * tree * tree

let is_valid_bst root =
  let rec go lo hi = function
    | Nil -> true
    | Node (v, l, r) ->
      v > lo && v < hi &&
      go lo v l && go v hi r
  in go min_int max_int root

File: solutions/scala/060-validate-binary-search-tree.scala
/**
 * Problem 60: Validate Binary Search Tree (LeetCode 98)
 * Difficulty: Med
 * Language: Scala
 */
sealed trait Tree[+A]
case object Nil extends Tree[Nothing]
case class Node[A](v: A, l: Tree[A], r: Tree[A]) extends Tree[A]

def isValidBST(t: Tree[Int], lo: Long = Long.MinValue, hi: Long = Long.MaxValue): Boolean =
  t match {
    case Nil => true
    case Node(v, l, r) =>
      v > lo && v < hi &&
      isValidBST(l, lo, v) &&
      isValidBST(r, v, hi)
  }
